File: Day_4/Sliding_Window/code/Solution_3/sol.java
package Day_4.Sliding_Window.code.Solution_3;

import java.util.*;

public class sol {
    public static String minWindow(String s, String t) {
        if (s == null || t == null || s.length() < t.length()) return "";

        Map<Character, Integer> dictT = new HashMap<>();
        for (char c : t.toCharArray()) {
            dictT.put(c, dictT.getOrDefault(c, 0) + 1);
        }
        int required = dictT.size();

        int left = 0, right = 0;
        int formed = 0;
        Map<Character, Integer> windowCounts = new HashMap<>();
        int[] ans = {-1, 0, 0}; // length, left, right

        while (right < s.length()) {
            char c = s.charAt(right);
            windowCounts.put(c, windowCounts.getOrDefault(c, 0) + 1);

            if (dictT.containsKey(c) && windowCounts.get(c).intValue() == dictT.get(c).intValue()) {
                formed++;
            }

            while (left <= right && formed == required) {
                if (ans[0] == -1 || right - left + 1 < ans[0]) {
                    ans[0] = right - left + 1;
                    ans[1] = left;
                    ans[2] = right;
                }

                char leftChar = s.charAt(left);
                windowCounts.put(leftChar, windowCounts.get(leftChar) - 1);
                if (dictT.containsKey(leftChar) && windowCounts.get(leftChar).intValue() < dictT.get(leftChar).intValue()) {
                    formed--;
                }
                left++;
            }

            right++;
        }

        return ans[0] == -1 ? "" : s.substring(ans[1], ans[2] + 1);
    }

    // Example driver
    public static void main(String[] args) {
        String s = "ADOBECODEBANC", t = "ABC";
        System.out.println(minWindow(s, t));  // Output: "BANC"
    }
}

File: Day_4/Sliding_Window/code/Solution_3/sol.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <climits>
using namespace std;

string minWindow(string s, string t) {
    if (s.empty() || t.empty()) return "";

    unordered_map<char, int> dict_t;
    for (char c : t) dict_t[c]++;

    int required = dict_t.size();
    unordered_map<char, int> window_counts;

    int left = 0, right = 0;
    int formed = 0;

    int min_len = INT_MAX, min_left = 0;

    while (right < s.size()) {
        char c = s[right];
        window_counts[c]++;

        if (dict_t.count(c) && window_counts[c] == dict_t[c]) {
            formed++;
        }

        while (left <= right && formed == required) {
            if (right - left + 1 < min_len) {
                min_len = right - left + 1;
                min_left = left;
            }

            char leftChar = s[left];
            window_counts[leftChar]--;
            if (dict_t.count(leftChar) && window_counts[leftChar] < dict_t[leftChar]) {
                formed--;
            }
            left++;
        }
        right++;
    }

    return min_len == INT_MAX ? "" : s.substr(min_left, min_len);
}

// Example driver
int main() {
    string s = "ADOBECODEBANC", t = "ABC";
    cout << minWindow(s, t) << endl;  // Output: "BANC"
    return 0;
}

